Upgrade of Browserify fails tests

* Sinon version : `master` (before updating) or branch `bug-updating-deps`
* Environment   : Chromium

**What did you expect to happen?**
All tests pass in all environments.

**What actually happens**
One test fails.

```
  1360 passing (919ms)
  4 pending
  1 failing

  1) fakeTimers.clock
       .useFakeTimers
         fakes provided methods:
     AssertionError: [refute.same] function onNextTick() {} expected not to be the same object as function onNextTick() {}
      at Object.fail (node_modules/@sinonjs/referee/lib/create-fail.js:5)
      at Object.fail (node_modules/@sinonjs/referee/lib/define-assertion.js:46)
      at assertion (node_modules/@sinonjs/referee/lib/define-assertion.js:68)
      at Function.referee.<computed>.<computed> [as same] (node_modules/@sinonjs/referee/lib/define-assertion.js:93)
      at Context.<anonymous> (test/util/fake-timers-test.js:1075)
```

**How to reproduce**

Check out `master` (671330ce). Run `npm update && npm test`.

Also pushed the updates: https://github.com/sinonjs/sinon/commit/c8ff2b8f6f438ccac7c5c3abd9a106dc84914610

As this bug manifested itself in the fake timers code, I assumed the bug was in Lolex, but alas, it turns out adding the same tests to lolex and upgrading Browserify (=>Chokidar=>fsevents=>process-nextick-args)  does _not_ manifest the bug. The same process-nextick-args version (2.0.0) was there (in Lolex), as well. Which points to this being the little code that is specific to Sinon's own `fake-timers.js`.  My guess in on the `module.exports` code.

<details>
<summary>output_pre_all_passing.txt</summary>

```
# chromium:


  fakeTimers.clock
    .setTimeout
      ✓ throws if no arguments
      ✓ returns numeric id or object with numeric id
      ✓ returns unique id
      ✓ sets timers on instance
      ✓ evals non-function callbacks
      ✓ passes setTimeout parameters
      ✓ calls correct timeout on recursive tick
    .setImmediate
      ✓ returns numeric id or object with numeric id
      ✓ calls the given callback immediately
      ✓ throws if no arguments
      ✓ manages separate timers per clock instance
      ✓ passes extra parameters through to the callback
    .clearImmediate
      ✓ removes immediate callbacks
    .tick
      ✓ triggers immediately without specified delay
      ✓ does not trigger without sufficient delay
      ✓ triggers after sufficient delay
      ✓ triggers simultaneous timers
      ✓ triggers multiple simultaneous timers
      ✓ triggers multiple simultaneous timers with zero callAt
      ✓ waits after setTimeout was called
      ✓ mini integration test
      ✓ triggers even when some throw
      ✓ calls function with global object or null (strict mode) as this
      ✓ triggers in the order scheduled
      ✓ creates updated Date while ticking
      ✓ fires timer in intervals of 13
      ✓ fires timers in correct order
      ✓ triggers timeouts and intervals in the order scheduled
      ✓ does not fire canceled intervals
      ✓ passes 6 seconds
      ✓ passes 1 minute
      ✓ passes 2 hours, 34 minutes and 12 seconds (63ms)
      ✓ throws for invalid format
      ✓ throws for invalid minutes
      ✓ throws for negative minutes
      ✓ treats missing argument as 0
      ✓ fires nested setTimeout calls properly
      ✓ does not silently catch exceptions
      ✓ returns the current now value
    .clearTimeout
      ✓ removes timeout
      ✓ ignores null argument
    .reset
      ✓ empties timeouts queue
    .setInterval
      ✓ throws if no arguments
      ✓ returns numeric id or object with numeric id
      ✓ returns unique id
      ✓ schedules recurring timeout
      ✓ does not schedule recurring timeout when cleared
      ✓ passes setTimeout parameters
    .date
      ✓ provides date constructor
      ✓ creates real Date objects
      ✓ creates real Date objects when called as function
      ✓ creates real Date objects when Date constructor is gone
      ✓ creates Date objects representing clock time
      ✓ returns Date object representing clock time
      ✓ listens to ticking clock
      ✓ creates regular date when passing timestamp
      ✓ returns regular date when calling with timestamp
      ✓ creates regular date when passing year, month
      ✓ returns regular date when calling with year, month
      ✓ creates regular date when passing y, m, d
      ✓ returns regular date when calling with y, m, d
      ✓ creates regular date when passing y, m, d, h
      ✓ returns regular date when calling with y, m, d, h
      ✓ creates regular date when passing y, m, d, h, m
      ✓ returns regular date when calling with y, m, d, h, m
      ✓ creates regular date when passing y, m, d, h, m, s
      ✓ returns regular date when calling with y, m, d, h, m, s
      ✓ creates regular date when passing y, m, d, h, m, s, ms
      ✓ returns regular date when calling with y, m, d, h, m, s, ms
      ✓ mirrors native Date.prototype
      ✓ supports now method if present
      ✓ mirrors parse method
      ✓ mirrors UTC method
      ✓ mirrors toUTCString method
      ✓ mirrors toString
      .now
        ✓ returns clock.now
      unsupported toSource
        ✓ is undefined
    .useFakeTimers
      ✓ returns clock object
      ✓ has clock property
      ✓ sets initial timestamp
      ✓ replaces global setTimeout
      ✓ global fake setTimeout should return id
      ✓ replaces global clearTimeout
      ✓ restores global setTimeout
      ✓ restores global clearTimeout
      ✓ replaces global setInterval
      ✓ replaces global clearInterval
      ✓ restores global setInterval
      ✓ restores global clearInterval
      ✓ deletes global property on restore if it was inherited onto the global object
      ✓ restores global property on restore if it is present on the global object itself
      ✓ fakes Date constructor
      ✓ fake Date constructor should mirror Date's properties
      ✓ decide on Date.now support at call-time when supported
      ✓ decide on Date.now support at call-time when unsupported
      ✓ mirrors custom Date properties
      ✓ restores Date constructor
setImmediate undefined
fakeTimers.timers.setImmediate undefined
setImmediate function () {
                return clock[method].apply(clock, arguments);
            }
fakeTimers.timers.setImmediate undefined
      ✓ fakes provided methods
      ✓ resets faked methods
      ✓ does not fake methods not provided
      ✓ installs by default without nextTick
      ✓ installs with nextTick
      ✓ installs clock in advancing mode and triggers setTimeout
      ✓ installs clock in advancing mode and triggers setInterval (40ms)
      ✓ installs clock in advancing mode and triggers setImmediate
      ✓ throws on old useFakeTimers signatures
      ✓ supports a way to pass the global context


  107 passing (377ms)
```

</details>


<details>
<summary>output_pre__grep_failing_setImmediate_undefined.txt</summary>

```
$ mochify test/util/fake-timers-test.js --grep "fakes provided methods"  > output_pre.txt

# chromium:


  fakeTimers.clock
    .useFakeTimers
      1) fakes provided methods


  0 passing (32ms)
  1 failing

  1) fakeTimers.clock
       .useFakeTimers
         fakes provided methods:
     ReferenceError: setImmediate is not defined
      at Context.<anonymous> (test/util/fake-timers-test.js:1072)
```
</details>

<details>
<summary>post_output.txt</summary>

```
$ cat output_post.txt 
# chromium:


  fakeTimers.clock
    .useFakeTimers
setImmediate function(fn) {
  var id = nextImmediateId++;
  var args = arguments.length < 2 ? false : slice.call(arguments, 1);

  immediateIds[id] = true;

  nextTick(function onNextTick() {
    if (immediateIds[id]) {
      // fn.call() is faster so we optimize for the common use-case
      // @see http://jsperf.com/call-apply-segu
      if (args) {
        fn.apply(null, args);
      } else {
        fn.call(null);
      }
      // Prevent ids from leaking
      exports.clearImmediate(id);
    }
  });

  return id;
}
fakeTimers.timers.setImmediate function(fn) {
  var id = nextImmediateId++;
  var args = arguments.length < 2 ? false : slice.call(arguments, 1);

  immediateIds[id] = true;

  nextTick(function onNextTick() {
    if (immediateIds[id]) {
      // fn.call() is faster so we optimize for the common use-case
      // @see http://jsperf.com/call-apply-segu
      if (args) {
        fn.apply(null, args);
      } else {
        fn.call(null);
      }
      // Prevent ids from leaking
      exports.clearImmediate(id);
    }
  });

  return id;
}
setImmediate function(fn) {
  var id = nextImmediateId++;
  var args = arguments.length < 2 ? false : slice.call(arguments, 1);

  immediateIds[id] = true;

  nextTick(function onNextTick() {
    if (immediateIds[id]) {
      // fn.call() is faster so we optimize for the common use-case
      // @see http://jsperf.com/call-apply-segu
      if (args) {
        fn.apply(null, args);
      } else {
        fn.call(null);
      }
      // Prevent ids from leaking
      exports.clearImmediate(id);
    }
  });

  return id;
}
fakeTimers.timers.setImmediate function(fn) {
  var id = nextImmediateId++;
  var args = arguments.length < 2 ? false : slice.call(arguments, 1);

  immediateIds[id] = true;

  nextTick(function onNextTick() {
    if (immediateIds[id]) {
      // fn.call() is faster so we optimize for the common use-case
      // @see http://jsperf.com/call-apply-segu
      if (args) {
        fn.apply(null, args);
      } else {
        fn.call(null);
      }
      // Prevent ids from leaking
      exports.clearImmediate(id);
    }
  });

  return id;
}
      1) fakes provided methods


  0 passing (79ms)
  1 failing

  1) fakeTimers.clock
       .useFakeTimers
         fakes provided methods:
     AssertionError: [refute.same] function onNextTick() {} expected not to be the same object as function onNextTick() {}
      at Object.fail (node_modules/@sinonjs/referee/lib/create-fail.js:5)
      at Object.fail (node_modules/@sinonjs/referee/lib/define-assertion.js:46)
      at assertion (node_modules/@sinonjs/referee/lib/define-assertion.js:68)
      at Function.referee.<computed>.<computed> [as same] (node_modules/@sinonjs/referee/lib/define-assertion.js:93)
      at Context.<anonymous> (test/util/fake-timers-test.js:1080)

```
</details>

--- a/lib/sinon/util/fake-timers.js
+++ b/lib/sinon/util/fake-timers.js
@@ -13,6 +13,13 @@ function createClock(config, globalCtx) {
     return clock;
 }
 
+function addIfDefined(obj, globalPropName) {
+    var globalProp = global[globalPropName];
+    if (typeof globalProp !== "undefined") {
+        obj[globalPropName] = globalProp;
+    }
+}
+
 /**
  * @param {number|Date|Object} dateOrConfig The unix epoch value to install with (default 0)
  * @returns {Object} Returns a lolex clock instance
@@ -51,12 +58,14 @@ exports.clock = {
     }
 };
 
-exports.timers = {
+var timers = {
     setTimeout: setTimeout,
     clearTimeout: clearTimeout,
-    setImmediate: typeof setImmediate !== "undefined" ? setImmediate : undefined,
-    clearImmediate: typeof clearImmediate !== "undefined" ? clearImmediate : undefined,
     setInterval: setInterval,
     clearInterval: clearInterval,
     Date: Date
 };
+addIfDefined(timers, "setImmediate");
+addIfDefined(timers, "clearImmediate");
+
+exports.timers = timers;

--- a/package.json
+++ b/package.json
@@ -58,7 +58,7 @@
     "@sinonjs/formatio": "^3.2.1",
     "@sinonjs/samsam": "^3.3.3",
     "diff": "^4.0.1",
-    "lolex": "^4.2.0",
+    "lolex": "^5.0.0",
     "nise": "^1.5.2",
     "supports-color": "^7.1.0"
   },

--- a/test/util/fake-timers-test.js
+++ b/test/util/fake-timers-test.js
@@ -102,56 +102,62 @@ describe("fakeTimers.clock", function() {
             this.clock = fakeTimers.clock.create();
         });
 
-        it("returns numeric id or object with numeric id", function() {
-            var result = this.clock.setImmediate(function() {
-                return;
+        if (typeof setImmediate === "function") {
+            it("returns numeric id or object with numeric id", function() {
+                var result = this.clock.setImmediate(function() {
+                    return;
+                });
+
+                if (typeof result === "object") {
+                    assert.isNumber(result.id);
+                } else {
+                    assert.isNumber(result);
+                }
             });
 
-            if (typeof result === "object") {
-                assert.isNumber(result.id);
-            } else {
-                assert.isNumber(result);
-            }
-        });
+            it("calls the given callback immediately", function() {
+                var stub = sinonStub();
 
-        it("calls the given callback immediately", function() {
-            var stub = sinonStub();
+                this.clock.setImmediate(stub);
+                this.clock.tick(0);
 
-            this.clock.setImmediate(stub);
-            this.clock.tick(0);
-
-            assert(stub.called);
-        });
-
-        it("throws if no arguments", function() {
-            var clock = this.clock;
-
-            assert.exception(function() {
-                clock.setImmediate();
+                assert(stub.called);
             });
-        });
 
-        it("manages separate timers per clock instance", function() {
-            var clock1 = fakeTimers.clock.create();
-            var clock2 = fakeTimers.clock.create();
-            var stubs = [sinonStub(), sinonStub()];
+            it("throws if no arguments", function() {
+                var clock = this.clock;
 
-            clock1.setImmediate(stubs[0]);
-            clock2.setImmediate(stubs[1]);
-            clock2.tick(0);
+                assert.exception(function() {
+                    clock.setImmediate();
+                });
+            });
 
-            assert.isFalse(stubs[0].called);
-            assert(stubs[1].called);
-        });
+            it("manages separate timers per clock instance", function() {
+                var clock1 = fakeTimers.clock.create();
+                var clock2 = fakeTimers.clock.create();
+                var stubs = [sinonStub(), sinonStub()];
 
-        it("passes extra parameters through to the callback", function() {
-            var stub = sinonStub();
+                clock1.setImmediate(stubs[0]);
+                clock2.setImmediate(stubs[1]);
+                clock2.tick(0);
 
-            this.clock.setImmediate(stub, "value1", 2);
-            this.clock.tick(1);
+                assert.isFalse(stubs[0].called);
+                assert(stubs[1].called);
+            });
 
-            assert(stub.calledWithExactly("value1", 2));
-        });
+            it("passes extra parameters through to the callback", function() {
+                var stub = sinonStub();
+
+                this.clock.setImmediate(stub, "value1", 2);
+                this.clock.tick(1);
+
+                assert(stub.calledWithExactly("value1", 2));
+            });
+        } else {
+            it("shouldn't install setImmedate", function() {
+                refute.isFunction(this.clock.setImmediate);
+            });
+        }
     });
 
     describe(".clearImmediate", function() {
@@ -159,15 +165,21 @@ describe("fakeTimers.clock", function() {
             this.clock = fakeTimers.clock.create();
         });
 
-        it("removes immediate callbacks", function() {
-            var callback = sinonStub();
+        if (typeof clearImmediate === "function") {
+            it("removes immediate callbacks", function() {
+                var callback = sinonStub();
 
-            var id = this.clock.setImmediate(callback);
-            this.clock.clearImmediate(id);
-            this.clock.tick(1);
+                var id = this.clock.setImmediate(callback);
+                this.clock.clearImmediate(id);
+                this.clock.tick(1);
 
-            assert.isFalse(callback.called);
-        });
+                assert.isFalse(callback.called);
+            });
+        } else {
+            it("shouldn't install clearImmedate", function() {
+                refute.isFunction(this.clock.clearImmediate);
+            });
+        }
     });
 
     describe(".tick", function() {
@@ -980,6 +992,42 @@ describe("fakeTimers.clock", function() {
 
             assert.same(clearInterval, fakeTimers.timers.clearInterval);
         });
+
+        if (typeof setImmediate === "function") {
+            it("restores global setImmediate", function() {
+                this.clock = fakeTimers.useFakeTimers();
+                var stub = sinonStub();
+                this.clock.restore();
+
+                this.timer = setImmediate(stub);
+                this.clock.tick(1);
+
+                assert.isFalse(stub.called);
+                assert.same(setImmediate, fakeTimers.timers.setImmediate);
+            });
+        } else {
+            it("does not install global setImmediate", function() {
+                this.clock = fakeTimers.useFakeTimers();
+
+                assert.isUndefined(setImmediate);
+            });
+        }
+
+        if (typeof clearImmediate === "function") {
+            it("restores global clearImmediate", function() {
+                this.clock = fakeTimers.useFakeTimers();
+                sinonStub();
+                this.clock.restore();
+
+                assert.same(clearImmediate, fakeTimers.timers.clearImmediate);
+            });
+        } else {
+            it("does not install global clearImmediate", function() {
+                this.clock = fakeTimers.useFakeTimers();
+
+                assert.isUndefined(clearImmediate);
+            });
+        }
 
         /*eslint-disable no-proto*/
         if (Object.__proto__) {

--- a/test/util/fake-timers-test.js
+++ b/test/util/fake-timers-test.js
@@ -644,10 +644,10 @@ describe("fakeTimers.clock", function() {
             assert(Date.prototype.isPrototypeOf(date));
         });
 
-        it("creates real Date objects when called as function", function() {
+        it("creates date strings when called as function", function() {
             var date = this.clock.Date();
 
-            assert(Date.prototype.isPrototypeOf(date));
+            assert.isString(date);
         });
 
         it("creates real Date objects when Date constructor is gone", function() {
@@ -676,12 +676,6 @@ describe("fakeTimers.clock", function() {
             assert.equals(date.getTime(), new Date(this.now).getTime());
         });
 
-        it("returns Date object representing clock time", function() {
-            var date = this.clock.Date();
-
-            assert.equals(date.getTime(), new Date(this.now).getTime());
-        });
-
         it("listens to ticking clock", function() {
             var date1 = new this.clock.Date();
             this.clock.tick(3);
@@ -697,23 +691,9 @@ describe("fakeTimers.clock", function() {
             assert.equals(fakeDate.getTime(), date.getTime());
         });
 
-        it("returns regular date when calling with timestamp", function() {
-            var date = new Date();
-            var fakeDate = this.clock.Date(date.getTime());
-
-            assert.equals(fakeDate.getTime(), date.getTime());
-        });
-
         it("creates regular date when passing year, month", function() {
             var date = new Date(2010, 4);
             var fakeDate = new this.clock.Date(2010, 4);
-
-            assert.equals(fakeDate.getTime(), date.getTime());
-        });
-
-        it("returns regular date when calling with year, month", function() {
-            var date = new Date(2010, 4);
-            var fakeDate = this.clock.Date(2010, 4);
 
             assert.equals(fakeDate.getTime(), date.getTime());
         });
@@ -725,23 +705,9 @@ describe("fakeTimers.clock", function() {
             assert.equals(fakeDate.getTime(), date.getTime());
         });
 
-        it("returns regular date when calling with y, m, d", function() {
-            var date = new Date(2010, 4, 2);
-            var fakeDate = this.clock.Date(2010, 4, 2);
-
-            assert.equals(fakeDate.getTime(), date.getTime());
-        });
-
         it("creates regular date when passing y, m, d, h", function() {
             var date = new Date(2010, 4, 2, 12);
             var fakeDate = new this.clock.Date(2010, 4, 2, 12);
-
-            assert.equals(fakeDate.getTime(), date.getTime());
-        });
-
-        it("returns regular date when calling with y, m, d, h", function() {
-            var date = new Date(2010, 4, 2, 12);
-            var fakeDate = this.clock.Date(2010, 4, 2, 12);
 
             assert.equals(fakeDate.getTime(), date.getTime());
         });
@@ -753,13 +719,6 @@ describe("fakeTimers.clock", function() {
             assert.equals(fakeDate.getTime(), date.getTime());
         });
 
-        it("returns regular date when calling with y, m, d, h, m", function() {
-            var date = new Date(2010, 4, 2, 12, 42);
-            var fakeDate = this.clock.Date(2010, 4, 2, 12, 42);
-
-            assert.equals(fakeDate.getTime(), date.getTime());
-        });
-
         it("creates regular date when passing y, m, d, h, m, s", function() {
             var date = new Date(2010, 4, 2, 12, 42, 53);
             var fakeDate = new this.clock.Date(2010, 4, 2, 12, 42, 53);
@@ -767,23 +726,9 @@ describe("fakeTimers.clock", function() {
             assert.equals(fakeDate.getTime(), date.getTime());
         });
 
-        it("returns regular date when calling with y, m, d, h, m, s", function() {
-            var date = new Date(2010, 4, 2, 12, 42, 53);
-            var fakeDate = this.clock.Date(2010, 4, 2, 12, 42, 53);
-
-            assert.equals(fakeDate.getTime(), date.getTime());
-        });
-
         it("creates regular date when passing y, m, d, h, m, s, ms", function() {
             var date = new Date(2010, 4, 2, 12, 42, 53, 498);
             var fakeDate = new this.clock.Date(2010, 4, 2, 12, 42, 53, 498);
-
-            assert.equals(fakeDate.getTime(), date.getTime());
-        });
-
-        it("returns regular date when calling with y, m, d, h, m, s, ms", function() {
-            var date = new Date(2010, 4, 2, 12, 42, 53, 498);
-            var fakeDate = this.clock.Date(2010, 4, 2, 12, 42, 53, 498);
 
             assert.equals(fakeDate.getTime(), date.getTime());
         });

--- a/test/util/fake-timers-test.js
+++ b/test/util/fake-timers-test.js
@@ -8,6 +8,7 @@ var sinonSpy = require("../../lib/sinon/spy");
 var assert = referee.assert;
 var refute = referee.refute;
 var GlobalDate = Date;
+var setImmediatePresent = global.setImmediate && typeof global.setImmediate === "function";
 
 describe("fakeTimers.clock", function() {
     beforeEach(function() {
@@ -1116,24 +1117,22 @@ describe("fakeTimers.clock", function() {
         });
 
         it("fakes provided methods", function() {
-            this.clock = fakeTimers.useFakeTimers({ toFake: ["setTimeout", "Date", "setImmediate"] });
+            this.clock = fakeTimers.useFakeTimers({ toFake: ["setTimeout", "Date"] });
 
             refute.same(setTimeout, fakeTimers.timers.setTimeout);
             refute.same(Date, fakeTimers.timers.Date);
-            refute.same(setImmediate, fakeTimers.timers.setImmediate);
         });
 
         it("resets faked methods", function() {
-            this.clock = fakeTimers.useFakeTimers({ toFake: ["setTimeout", "Date", "setImmediate"] });
+            this.clock = fakeTimers.useFakeTimers({ toFake: ["setTimeout", "Date"] });
             this.clock.restore();
 
             assert.same(setTimeout, fakeTimers.timers.setTimeout);
             assert.same(Date, fakeTimers.timers.Date);
-            assert.same(setImmediate, fakeTimers.timers.setImmediate);
         });
 
         it("does not fake methods not provided", function() {
-            this.clock = fakeTimers.useFakeTimers({ toFake: ["setTimeout", "Date", "setImmediate"] });
+            this.clock = fakeTimers.useFakeTimers({ toFake: ["setTimeout", "Date"] });
 
             assert.same(clearTimeout, fakeTimers.timers.clearTimeout);
             assert.same(setInterval, fakeTimers.timers.setInterval);
@@ -1191,6 +1190,10 @@ describe("fakeTimers.clock", function() {
         });
 
         it("installs clock in advancing mode and triggers setImmediate", function(done) {
+            if (!setImmediatePresent) {
+                this.skip();
+            }
+
             this.clock = fakeTimers.useFakeTimers({ shouldAdvanceTime: true });
             this.clock.setImmediate(
                 function() {

